### ISSUE TITLE
Sequence tagging custom dataset

### ISSUE THREAD

**Issue by antgr** [pytorch/text#658]
## ❓ Questions and Help

**Description**

Hi, I have a custom dataset that has the following format:
```
Word1	O	O	N	s: 1	Sentence: 1	Doc: 1
Word2	O	O	N	s: 1	Sentence: 1	Doc: 1
Word3	O	O	N	s: 1	Sentence: 1	Doc: 1
Word4	O	O	N	s: 1	Sentence: 1	Doc: 1
```
I want to use column 0 as my sentences, and the next three consecutive columns as my labels (label1, label2, label3). I could afford to ignore the other fields. (Maybe in the future I would consider to use the last column, for example I have an idea to zero the gradient only when I switch document, and not in switching sentence, and I would like to test it, if that makes sense).

Could you help me on how I could read this dataset? 
For example to point me out a similar example in the documentation.
Thank you for your support!

**Comment by zhangguanheng66** [comment]
@mttk does TabularDataset work for this?

You could also a custom pipeline to read the text file. Some reference https://github.com/pytorch/text/pull/651

**Comment by mttk** [comment]
For this case, you should subclass the [SequenceTaggingDataset](https://github.com/pytorch/text/blob/master/torchtext/datasets/sequence_tagging.py#L5). The examples in your dataset should be in the format used by, for example, UDPOS (a newline separating instances). I think this is already fine in your case. You have examples of subclassing in the same file.

For an example of Field definition for dataset loading, you can check [this](https://github.com/pytorch/text/blob/master/test/sequence_tagging.py#L10) file.

**Comment by antgr** [comment]
OK thanks! I will look into your suggestions with detail! The fact that I have the train, dev, test, already separated, makes any difference?

**Comment by mttk** [comment]
It should even make the process simpler. Ideally, you should just implement the `splits` method in the SequenceTaggingDatset subclass, pass the correct paths to the files, and everything should work out of the box.

**Comment by antgr** [comment]
What I have done following the answer and the test :
*(I have changed the specific names to lab1,lab2,lab3 for more generality)*
```
class MySeqLabelDS(SequenceTaggingDataset):
    @classmethod
    def splits(cls, fields, root=".data", train="train.txt",
               validation="dev.txt",
               test="test.txt", **kwargs):

        return super(MySeqLabelDS, cls).splits(
            fields=fields, root=root, train=train, validation=validation,
            test=test, **kwargs)
```
and
```
WORD = data.Field(init_token="<bos>", eos_token="<eos>")
LAB1 = data.Field(init_token="<bos>", eos_token="<eos>")
LAB2 = data.Field(init_token="<bos>", eos_token="<eos>")
LAB3 = data.Field(init_token="<bos>", eos_token="<eos>")
```
and finaly:
```
train, val, test = MySeqLabelDS.splits(
    fields=(('word', WORD), ('lab1', LAB1), ('lab2', LAB2), ('lab3', LAB3)),
    path="./datasetl",
    train="train.txt",
    validation="dev.txt",
    test="test.txt")

WORD.build_vocab(train.word, min_freq=3)
LAB1.build_vocab(train.lab1)
LAB2.build_vocab(train.lab2)
LAB3.build_vocab(train.lab3)

train_iter, val_iter = data.BucketIterator.splits(
    (train, val), batch_size=3, device="cuda:0")
```

**Comment by mttk** [comment]
I think that is the convention so you would get batch elements (.word, .lab1) of the same length, but it is not necessary (you will have to handle that yourself if you choose not to assign them).
You can (should) mask them out at prediction time as they don't really contribute to the loss.

**Comment by antgr** [comment]
How does this (init_token and eos_token) help get batch elements (.word, .lab1) of the same length? 
How could I handle that by myself? (What I am supposed to do?). I miss the connection.

If I understand correctly, I always have to mask them out in forward and in computation of loss, right? For example to take ```logits[1:-1]``` ?
 
Thank you!

**Comment by mttk** [comment]
I assumed that you want to have `init_token` and `eos_token` in your WORD Field. In that case, the length of an `batch.word` will be longer by 2 than `batch.lab1` (if you didn't set those tokens there as well). I'm guessing the example adds init and eos tokens to labels for convenience.

Doing `logits[1:-1]` is a correct way to handle this (in that case, your `labels` don't need to have init and eos).
Sorry for the confusion, glad it works!

**Comment by antgr** [comment]
OK, it makes sense. So, if I will use ```init_token``` and ```eos_token``` for labels as well, I have to do ```labels[1:-1]``` as well. Is that correct?  Is this the standard way for someone using torchtext, to use a filtering such as ```[1:-1]``` for logits and labels? There is no confusion by your explanations!  The confusion is probably due to my lack of familiarity! Thanks!

**Comment by mttk** [comment]
Yeah. Although I am pretty sure that just doing `logits[1:-1]` and not setting init and eos tokens for labels will work.
I'm honestly confused right now why the labels also have init and eos tokens, since I believe you should never predict those.
